packages/backend/server/src/plugins/copilot/session.ts

```diff
@@ -0,0 +1,216 @@
+import { randomUUID } from 'node:crypto';
+
+import { Injectable, Logger } from '@nestjs/common';
+import { PrismaClient } from '@prisma/client';
+import { encoding_for_model, Tiktoken, TiktokenModel } from 'tiktoken';
+
+import { SessionCache } from '../../fundamentals';
+import { PromptService } from './prompt';
+import { ChatMessage, ChatMessageSchema } from './types';
+
+const CHAT_SESSION_KEY = 'chat-session';
+const CHAT_SESSION_TTL = 3600 * 12 * 1000; // 12 hours
+
+export interface ChatSessionOptions {
+  // connect ids
+  userId: string;
+  workspaceId: string;
+  docId: string;
+  promptName: string;
+  // options
+  action: boolean;
+  model: TiktokenModel;
+}
+
+export interface ChatSessionState extends ChatSessionOptions {
+  // connect ids
+  sessionId: string;
+  // states
+  prompt: ChatMessage[];
+  messages: ChatMessage[];
+}
+
+export class ChatSession implements AsyncDisposable {
+  private readonly encoder: Tiktoken;
+  private readonly promptTokenSize: number;
+  constructor(
+    private readonly state: ChatSessionState,
+    model: TiktokenModel,
+    private readonly dispose?: (state: ChatSessionState) => Promise<void>,
+    private readonly maxTokenSize = 3840
+  ) {
+    this.encoder = encoding_for_model(model);
+    this.promptTokenSize = this.encoder.encode_ordinary(
+      state.prompt?.map(m => m.content).join('') || ''
+    ).length;
+  }
+
+  get model() {
+    return this.state.model;
+  }
+
+  push(message: ChatMessage) {
+    this.state.messages.push(message);
+  }
+
+  pop() {
+    this.state.messages.pop();
+  }
+
+  private takeMessages(): ChatMessage[] {
+    if (this.state.action) {
+      const messages = this.state.messages;
+      return messages.slice(messages.length - 1);
+    }
+    const ret = [];
+    const messages = [...this.state.messages];
+    messages.reverse();
```

Comment:
why reverse?

Reply:
used to calculate message token to keep as much of the chat history as possible

Reply:
iterate messages from end to front will save one reverse turn

Reply:
<img width="892" alt="image" src="https://github.com/toeverything/AFFiNE/assets/25152247/32732802-9cc2-408a-83a0-cb1e51e6d92c">

just make bench and found that reverse would indeed cause a significant performance decline.
I have switched to while loop